### PR TITLE
All remaining subcommands for the h3 cli program

### PR DESCRIPTION
**Pull request by dfellis**
After implementing the region subcommands, all of the rest documented on [h3geo.org](https://h3geo.org) were "trivial" as all patterns of subcommand had been implemented, so I just slogged through writing *all* of them while I had the drive to do it. ;)

There may be some test failures with `degsToRads` and friends due to floating point rounding errors on different CPU architectures. I hope not, but if so, then I'll have to figure out how to test them better.

- **areNeighborCells**
- **cellsToDirectedEdge**
- **isValidDirectedEdge**
- **getDirectedEdgeOrigin**
- **getDirectedEdgeDestination**
- **directedEdgeToCells**
- **originToDirectedEdges**
- **directedEdgeToBoundary**
- **cellToVertex**
- **cellToVertexes**
- **vertexToLatLng**
- **isValidVertex**
- **degsToRads radsToDegs**
- **getHexagonAreaAvgKm2**
- **getHexagonAreaAvgM2**
- **cellAreaRads2**
- **cellAreaKm2**
- **cellAreaM2**
- **getHexagonEdgeLengthAvgKm**
- **getHexagonEdgeLengthAvgM**
- **edgeLengthKm**
- **edgeLengthM**
- **edgeLengthRads**
- **getNumCells**
- **getRes0Cells**
- **getPentagons**
- **pentagonCount**
- **greatCircleDistanceKm**
- **greatCircleDistanceM**
- **greatCircleDistanceRads**
- **describeH3Error**

### CODE REVIEW

**Comment by dfellis** [comment]
Oh, this currently includes the region subcommands in this PR, since it depends on that one, but once it gets approved and merged, this'll be reduced to only the new commands.

I did (almost) every command in a separate commit, for easier reviewing.

File: src/apps/filters/h3.c [hunk 1]
```diff
@@ -1189,6 +1189,1238 @@ SUBCOMMAND(uncompactCells,
...
+                  .value = &res,
+                  .helpText = "Resolution, 0-15 inclusive."};
+    Arg *args[] = {&getNumCellsArg, &resArg, &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    int64_t numCells = 0;
+    H3Error err = H3_EXPORT(getNumCells)(res, &numCells);
+    if (err) {
+        return err;
+    }
+    printf("%" PRIu64 "\n", numCells);
```
Comment (isaacbrodsky) [inline src/apps/filters/h3.c:2126]:
```suggestion
    printf("%" PRId64 "\n", numCells);
```
nit, it's a signed integer

File: src/apps/filters/h3.c [hunk 2]
```diff
@@ -1189,6 +1189,1238 @@ SUBCOMMAND(uncompactCells,
...
+        .names = {"-c", "--coordinates"},
+        .scanFormat = "%1500c",
+        .valueName = "ARRAY",
+        .value = &coordinateStr,
+        .helpText =
+            "The array of coordinates to convert. Up to 1500 characters."};
+    Arg *args[] = {&greatCircleDistanceRadsArg, &filenameArg, &coordinateStrArg,
+                   &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    if (!filenameArg.found && !coordinateStrArg.found) {
```
Comment (isaacbrodsky) [inline src/apps/filters/h3.c:2217]:
```suggestion
    if (filenameArg.found == coordinateStrArg.found) {
```

both shouldn't be provided, right?

File: src/apps/filters/h3.c [hunk 3]
```diff
@@ -1189,6 +1189,1238 @@ SUBCOMMAND(uncompactCells,
...
+        .names = {"-c", "--coordinates"},
+        .scanFormat = "%1500c",
+        .valueName = "ARRAY",
+        .value = &coordinateStr,
+        .helpText =
+            "The array of coordinates to convert. Up to 1500 characters."};
+    Arg *args[] = {&greatCircleDistanceKmArg, &filenameArg, &coordinateStrArg,
+                   &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    if (!filenameArg.found && !coordinateStrArg.found) {
```
Comment (isaacbrodsky) [inline src/apps/filters/h3.c:2288]:
```suggestion
    if (filenameArg.found == coordinateStrArg.found) {
```

both shouldn't be provided, right?

File: src/apps/filters/h3.c [hunk 4]
```diff
@@ -1189,6 +1189,1238 @@ SUBCOMMAND(uncompactCells,
...
+        .names = {"-c", "--coordinates"},
+        .scanFormat = "%1500c",
+        .valueName = "ARRAY",
+        .value = &coordinateStr,
+        .helpText =
+            "The array of coordinates to convert. Up to 1500 characters."};
+    Arg *args[] = {&greatCircleDistanceMArg, &filenameArg, &coordinateStrArg,
+                   &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    if (!filenameArg.found && !coordinateStrArg.found) {
```
Comment (isaacbrodsky) [inline src/apps/filters/h3.c:2359]:
```suggestion
    if (filenameArg.found == coordinateStrArg.found) {
```

as above

**Review by isaacbrodsky — COMMENTED** [review]
This looks good to me, pending #923 merging in

File: src/apps/filters/h3.c (hunk 2 above)
Comment (dfellis) [reply src/apps/filters/h3.c:2217]: As mentioned in the prior PR, no. Either a file is provided or the coordinates are provided in-line in the call, not both.

File: src/apps/filters/h3.c (hunk 3 above)
Comment (dfellis) [reply src/apps/filters/h3.c:2288]: Same above, this change is incorrect.

File: src/apps/filters/h3.c (hunk 4 above)
Comment (dfellis) [reply src/apps/filters/h3.c:2359]: Same as above, these arguments are alternatives of each other.

File: src/apps/filters/h3.c [hunk 5]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+    free(out);
+    return E_SUCCESS;
+}
+
+SUBCOMMAND(pentagonCount, "Returns 12") {
+    Arg *args[] = {&pentagonCountArg, &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    printf("12\n");
+    return E_SUCCESS;
+}
```
Comment (isaacbrodsky) [inline src/apps/filters/h3.c:2209]: not sure if we really need to bind this, but ok

**Review by isaacbrodsky — APPROVED** [review]

File: src/apps/filters/h3.c (hunk 5 above)
Comment (dfellis) [reply src/apps/filters/h3.c:2209]: I personally find it hilarious, so I want it in. ;)

File: src/apps/filters/h3.c [hunk 6]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+
+SUBCOMMAND(isValidDirectedEdge,
+           "Checks if the provided H3 directed edge is actually valid") {
+    DEFINE_CELL_ARG(cell, cellArg);
+    Arg *args[] = {&isValidDirectedEdgeArg, &helpArg, &cellArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    bool isValid = H3_EXPORT(isValidDirectedEdge)(cell);
+    printf("%s", isValid ? "true\n" : "false\n");
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:1700]: No strong opinion here, but would `1` and `0` be more flexible outputs?

File: src/apps/filters/h3.c [hunk 7]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+    DEFINE_CELL_ARG(cell, cellArg);
+    Arg *args[] = {&directedEdgeToBoundaryArg, &cellArg, &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    CellBoundary cb = {0};
+    H3Error err = H3_EXPORT(directedEdgeToBoundary)(cell, &cb);
+    if (err) {
+        return err;
+    }
+    // Using WKT formatting for the output. TODO: Add support for JSON
+    // formatting
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:1789]: Is it weird that we use `WKT` here but raw `lat,lng` polygons for `cellsToMultiPolygon`?

File: src/apps/filters/h3.c [hunk 8]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+    bool isValid = H3_EXPORT(isValidCell)(cell);
+    if (!isValid) {
+        return E_CELL_INVALID;
+    }
+    H3Index out = 0;
+    H3Error err = H3_EXPORT(cellToVertex)(cell, vertIndex, &out);
+    if (err) {
+        return err;
+    }
+    printf("%" PRIx64 "\n", out);
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:1828]: Nit: Should we use `h3Println` here?

File: src/apps/filters/h3.c [hunk 9]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+    double rad = 0;
+    Arg radArg = {.names = {"-r", "--radian"},
+                  .required = true,
+                  .scanFormat = "%lf",
+                  .valueName = "RAD",
+                  .value = &rad,
+                  .helpText = "Angle in radians"};
+    Arg *args[] = {&radsToDegsArg, &radArg, &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    printf("%.10lf\n", H3_EXPORT(radsToDegs)(rad));
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:1919]: Nit: Worth making a common `printDouble` function to help us always print with the same format?

File: src/apps/filters/h3.c [hunk 10]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+    for (int i = 0; i < 122; i++) {
+        if (out[i] > 0) {
+            if (hasPrinted) {
+                printf(", ");
+            }
+            printf("%" PRIx64, out[i]);
+            hasPrinted = true;
+        }
+    }
+    printf("]\n");
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:2166]: For `polygonToCells` we print newline-delimited cells, but most of the multi-cell outputs here use JSON arrays. Should this be consistent?

File: src/apps/filters/h3.c [hunk 11]
```diff
@@ -1632,6 +1632,811 @@ SUBCOMMAND(cellsToMultiPolygon,
...
+                   &helpArg};
+    PARSE_SUBCOMMAND(argc, argv, args);
+    if (filenameArg.found == coordinateStrArg.found) {
+        fprintf(
+            stderr,
+            "You must provide either a file to read from or a coordinate array "
+            "to use greatCircleDistanceKm");
+        exit(1);
+    }
+    FILE *fp = 0;
```
Comment (nrabinowitz) [inline src/apps/filters/h3.c:2311]: Nit: It feels like we could share a lot of code between `greatCircleDistanceKm` and `greatCircleDistanceRads` and `greatCircleDistanceM`

**Review by nrabinowitz — APPROVED** [review]
A few questions and nits, but otherwise looks great!

**Comment by dfellis** [comment]
Discussed offline, some of the inconsistencies will be taken care of in a follow-up PR.

**Pull request merged**